### ISSUE TITLE
Maintain line and character number after parsing

### ISSUE THREAD

**Issue by gregorybchris** [tekknolagi/scrapscript#86]
Evaluation has many failure cases, many of which have pretty reasonable error messages. However we do not maintain the line and character numbers associated with nodes in the AST. The interpreter can't point to the part of the code associated with a failure that occurs during evaluation.

This feature request is to make code index information available in the evaluation stage.

**Comment by tekknolagi** [comment]
This is a little finicky right now unless we want to store (invalid) line number information on all objects because the AST structures are objects themselves. Perhaps we should separate them out (ehhhh) or find another way to do a mixin or something.

**Comment by tekknolagi** [comment]
![Screenshot from 2024-01-10 11-31-29](https://github.com/tekknolagi/scrapscript/assets/401167/d86655cf-64c0-476c-b739-806445020f6a)

came across this out of nowhere this morning

**Comment by tekknolagi** [comment]
the reasoning is a little weird but i think we can calculate the line/col from the byte position easily and it's only one number to store

**Comment by tekknolagi** [comment]
I saw/skimmed part of a cool talk that did this for Circle, I think. Perhaps this talk https://www.youtube.com/watch?v=1m_5SVmGA4k

They have some cool tricks

**Comment by tekknolagi** [comment]
Oh, no, lmao, it was Carbon: https://www.youtube.com/watch?v=ZI198eFghJk
Starting ~30mins?

**Comment by tekknolagi** [comment]
cc @neuroevolutus maybe this is also interesting

I think we want to keep line/col on all AST objects but not on interstitial/value objects. This probably means setting them dynamically after construction and having helper functions `has_sourcepos` and `sourcepos`. Then hopefully we can use this in parse/eval/compile errors!

Maybe we can one day even emit `#line` directives in the compiler.

**Comment by neuroevolutus** [comment]
Neat, I'll have to take a look at the referenced article and video.

**Comment by neuroevolutus** [comment]
@tekknolagi I started working on a pull request for this, and I realised that adding additional fields to the `Token` class to represent column numbers and/or byte positions would cause the golden tests that check for the string contents of `ParseError`s to fail. Would it be acceptable to change those tests to instead test for the presence of the relevant fields and their specific values instead so that they become more immune to changes in the string representation of `ParseError`s?

**Comment by tekknolagi** [comment]
Sure! Please split that out into a separate commit (same PR) onto which you stack your lineno commit.

**Comment by tekknolagi** [comment]
In general, tests are as mutable as the rest of the code, and I like your philosophy of having stable tests. 

Some tests should check end-to-end string stuff though so that we have tests for stuff users see.

**Comment by neuroevolutus** [comment]
Although my goal is to keep progressing on this issue, I was mulling about whether we could split `scrapscript.py` into separate modules so that it does not keep growing larger and become harder to wrangle. I was thinking we could have separate modules for `lexer`, `parser`, `type_checker`, `optimizer` and `code_generator`, for example. Maybe they could all be children of the `scrapscript` module. Then, the tests could perhaps be grouped with their corresponding aforementioned submodules or be put into a `test` folder that roughly mirrors the structure of the `scrapscript` module. What are your thoughts on a change like this?

**Comment by tekknolagi** [comment]
I am probably less interested in splitting, except for splitting out tests (which are the bulk of the file) into `scrapscript_tests.py` or similar. I think as soon as we start thinking about the Python module system, we've lost :/

**Comment by neuroevolutus** [comment]
Haha, I can understand the sentiment. Having a separate `scrapscript_tests.py` would be a nice compromise. I would be glad to make a small PR for that if acceptable.

**Comment by tekknolagi** [comment]
Please feel free! Maybe leave the compiler tests where they are for now. 

We can also get rid of the `test` sub command

**Comment by neuroevolutus** [comment]
I decided to bite the bullet and avoid splitting out the tests into its own file for now.

I would be on board to remove the `test` subcommand. I think I can do that as well as update the documentation appropriately in another PR.

Right now, I think I've made all the necessary logical changes needed to keep the line number information during parsing. I just need to go through and manually fix the test failures that are occurring due to the AST nodes now having a `source_extent` field of type `Optional[SourceExtent]`.

**Comment by tekknolagi** [comment]
> I decided to bite the bullet and avoid splitting out the tests into its own file for now.

Big File remains undefeated
 
> I would be on board to remove the `test` subcommand. I think I can do that as well as update the documentation appropriately in another PR.

I think this only makes sense if tests have been split out into their own file

> Right now, I think I've made all the necessary logical changes needed to keep the line number information during parsing. I just need to go through and manually fix the test failures that are occurring due to the AST nodes now having a `source_extent` field of type `Optional[SourceExtent]`.

Nice! What are the failures? Tests expect lineno of -1 or something?

**Comment by neuroevolutus** [comment]
> Big File remains undefeated

Lol, true 🤣.

> I think this only makes sense if tests have been split out into their own file

I see; fair enough.

> Nice! What are the failures? Tests expect lineno of -1 or something?

Yeah, that's pretty much what's happening. Well, at least that seems to be the case for tests that start with a stream of tokens rather than with raw source text. For the tests that start with raw source text, the assertions fail because the parsed AST nodes are now actually maintaining meaningful line, column and byte numbers. I don't think there are any failures that are not straightforward like this, but there are a lot of them. I'm getting 145 total failures, so it will take a while to go through them 😅.

**Comment by tekknolagi** [comment]
I think it's fine for AST nodes to not all have source position attached; some won't because ASTs and objects live in the same object space. It shouldn't be an error to not have source position. So the tests should be fine as is

**Comment by neuroevolutus** [comment]
> I think it's fine for AST nodes to not all have source position attached; some won't because ASTs and objects live in the same object space. It shouldn't be an error to not have source position. So the tests should be fine as is

Yes, you are totally right, :facepalm:. I forgot to set the `compare` parameter of the `dataclasses.field` constructor to `False`.

**Comment by tekknolagi** [comment]
Nice, the bulk of the work has been done here now! We should probably think about designing an error reporting system.
